server/keyshare/tasks/tasks.go

```diff
@@ -159,28 +209,85 @@ func (t *taskHandler) expireAccounts(ctx context.Context) {
 			}
 
 			// Send emails
-			err = t.sendExpiryEmails(ctx, id, username, lang)
-			// FIXME: 'return nil' will prevent abortion of 'expireAccounts()'
-			// but will not take care of the actual problem of handling the invalid email
-			if err == keyshare.ErrInvalidEmail {
-				return nil
-			}
-			if err != nil {
-				return err // already logged, just abort
+			if err := t.sendExpiryEmails(ctx, id, username, lang); err != nil {
+
+				// To have the exact same behavior as before email revalidation functionality,
+				// we return nil when the error is ErrInvalidEmail
+				if !t.revalidateMail && err == keyshare.ErrInvalidEmail {
+					return nil
+				} else {
+					return err // already logged, just abort
+				}
 			}
 
 			// Finally, do marking for deletion
-			err = t.db.ExecUserContext(ctx, "UPDATE irma.users SET delete_on = $2 WHERE id = $1", id,
-				time.Now().Add(time.Duration(24*t.conf.DeleteDelay)*time.Hour).Unix())
-			if err != nil {
+			if err := t.db.ExecUserContext(ctx, "UPDATE irma.users SET delete_on = $2 WHERE id = $1",
+				id,
+				time.Now().Add(time.Duration(24*t.conf.DeleteDelay)*time.Hour).Unix()); err != nil {
 				return err
 			}
+
 			return nil
 		},
 		time.Now().Add(time.Duration(-24*t.conf.ExpiryDelay)*time.Hour).Unix(),
 	)
 	if err != nil {
-		t.conf.Logger.WithField("error", err).Error("Could not query for accounts that have expired")
+		t.conf.Logger.WithField("error", err).Error("could not query for accounts that have expired")
+		return
+	}
+}
+
+// When email revalidation is enabled, this task revalidates email addresses which where
+// flagged in expireAccounts due to being (temporary) invalid.
+func (t *taskHandler) revalidateMails(ctx context.Context) {
+
+	if !t.revalidateMail {
+		t.conf.Logger.Warning("email address revalidation is disabled because the emails.revalidate_on column is not present in the schema")
+		return
+	}
+
+	// Select only 100 records to prevent a potential storm of DNS requests
+	err := t.db.QueryIterateContext(ctx, `
+		SELECT e.id, e.email
+		FROM irma.emails AS e
+		WHERE e.revalidate_on < $1 
+		LIMIT 100`,
+		func(res *sql.Rows) error {
+			var id int64
+			var email string
+			err := res.Scan(&id, &email)
+			if err != nil {
+				return err
+			}
+
+			addr, err := keyshare.ParseEmailAddress(email)
+
+			if err != nil {
+				return err
+			}
+
+			if err := keyshare.VerifyMXRecord(addr.Host); err != nil {
```

Comment:
Would be nice, but don't think that'll work in here because validation of the email address should be done while constructing the object with the given email address string as input. So I do not see a way to do this without adding complexity. Do you? If so: please enlighten me:)

This is what I come up with:
```
type EmailAddress struct {
	Address string
	Host    string
}

func (e EmailAddress) Validate() error {
	addr, err := mail.ParseAddress(t.Address)
	if err != nil {
		return ErrInvalidEmail
	}

	e.Address = addr.Address
	e.Host = addr.Address[strings.LastIndex(addr.Address, "@")+1:]

	return nil
}

// and then:
	addr := EmailAddress{ Address: "email@domain.com" }
	
	if !addr.Validate() {
		...
	}
```

But this allows inconsistencies at Address and Host (can be overwritten/altered) and Validate (which sets the Host) is not forced so could be forgotten.

Also wanted to keep it's behavior the same as mail.ParseAddress(addr)

What do you think?

Reply:
It's not so practical to make changes inside `Validate` indeed. Then you'd need to parse the email address twice, once to create the struct and once to validate it. That might be a bit overkill indeed.